### DISCUSSION TITLE
Request providing a API to generate lockfile for offline build.

### DISCUSSION THREAD

**Discussion by sequencer** [com-lihaoyi/mill#3128] [Ideas]
In many build flows, reproducibility and offline building is important. 

I wonder if it would be possible to provide a tool from mill to expose all ivy as json as the lock file. Later we can use this file to download all ivy dependencies w/ coursier to prepare the build artifacts for no network building.

A dirty hack has been achieved in our nix flow:
https://github.com/chipsalliance/t1/blob/master/nix/pkgs/mill-builder.nix

**Comment by lefou** [comment]
Technically, all `ivyDeps` targets and friends (`runIvyDeps`, `compileIvyDeps`)  are persisted as json in the Mill cache after a first run. But this is probably not what you had in mind.

So, let me ask, do you already have some specific lock file format in mind, which we should produce? Is there already some lock file format which coursier or nix understands? If not, it would be helpful to gather some information about what we should put in such a file.

The lockfile generator could be implemented as a external module, since all targets providing ivyDeps are more or less introspectable. (The `contrib` space would be a suitable playground for it.) Once, we have some well established file format, we can think about integrating it directly into the modules that need it (`ScalaModule`, `JavaModule`, ...). At that stage, we should even be able fetch/use dependencies directly from persistent lockfiles. This would allow building in other sandbox-like build environments like e.g. portage / Gentoo Linux as well.

**Comment by sequencer** [comment]
May relate to https://github.com/coursier/coursier/issues/1487

**Reply by lefou** [reply]
There are two approaches possible:

1. Implement a reasonable serialization format for resolution in coursier, which can be used to fetch and validate later resolutions.
2. Implement the same in Mill directly

Given the comments in https://github.com/coursier/coursier/issues/1487, many parties from different build tools are interested in that feature, and they all have the same motivation for it, so I think approach 1, implementing it in coursier, would be the way forward. OTOH, coursier didn't attract many contributors, so it might be a bit harder to implement it there.

**Comment by sequencer** [comment]
I came across this issue again: ivy hash changed silently after one or two month and invalidates our caches....

> So, let me ask, do you already have some specific lock file format in mind, which we should produce?

> adding something like this might be enough:
```json
[
{
  "coord": "org.scala-lang:scala-library:2.13.1",
  "url": "https://repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.1/scala-library-2.13.1.jar",
  "sha256sum": "711dcb0ba3d8d8392a52162c1f251d89787148981d2823e30697bc08282521ae",
},
]
```

I prefer your second approach, "Implement the same in Mill directly", my concerns are:
- mill has multiple phase of downloading the dependencies, including downloading the external dependencies for `build.mill`, it should also be stored in this json dependencies file.
- In order to not change user build script, maybe the best solution is implement this in the mill itself rather than the `contrib` package.
- Mill should provide a offline mode to only resolve dependencies from only one local folder, I think this should be done with a environment variable.

In my mind, I think this might be possible:
- We can use a non-sandbox config to download and lock everything firstly, then give out a lock file in a json file.
- After this, an external tool or script should resolve this json to construct a local ivy folder, this is a tool specific, we can purely curl from the url rather than coursier to avoid tool specific problems.
- At the sandbox stage(maybe by `MILL_SANDBOXING` env), the local ivy folder should be provided to mill, and mill should only resolve dependencies from there.
